cuML/src/common/device_buffer.hpp

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "../../../ml-prims/src/utils.h"
+
+#include "../cuML.hpp"
+
+namespace ML {
+
+/**
+ * RAII object owning a contigous typed device buffer. The passed in allocator supports asynchronus allocation and
+ * deallocation so this can be used for temporary memory 
+ * @code{.cpp}
+ * void foo( cumlHandle* handle, .., cudaStream_t stream )
+ * {
+ *     ...
+ *     device_buffer<T> temp( handle->getHostAllocator(), 0 )
+ *     
+ *     temp.resize(n, stream);
+ *     kernelA<<<grid,block,0,stream>>>(...,temp.data(),...);
+ *     kernelB<<<grid,block,0,stream>>>(...,temp.data(),...);
+ *     temp.release(stream);
+ * }
+ * @endcode
+ */
+template<typename T>
+class device_buffer
+{
+public:
+    typedef std::size_t         size_type;
+    typedef T                   value_type;
+    typedef value_type*         iterator;
+    typedef const value_type*   const_iterator;
+    typedef T&                  reference;
+    typedef const T&            const_reference;
+
+    device_buffer() = delete;
+
+    device_buffer(const device_buffer& other) = delete;
+
+    device_buffer& operator=(const device_buffer& other) = delete;
+
+    device_buffer(std::shared_ptr<deviceAllocator> allocator, size_type n = 0)
+        : _allocator(allocator), _size(n), _capacity(n), _data(0)
+    {
+        if ( n > 0 )
+        {
+            _data = static_cast<value_type*>(_allocator->allocate( _capacity*sizeof(value_type), 0 ));
+            CUDA_CHECK( cudaStreamSynchronize( 0 ) );
```

Comment:
I'm a bit confused here. So effectively we are sync'ing on the default stream everytime we create a `device_buffer`?

Reply:
Yes, because there is no other stream available in the constructor to pass into the allocator and to synchronize on. I have chosen this approach because I think it would be confusing if a constructor is asynchronous. I would expect that an object is fully usable after construction. If a user provided stream should be used for the allocation and the synchronization should be avoided the technique described in the developer guide using `resize` and `release`  could be used. One alternative I could think is forcing users of device_buffer and host_buffer to use resize by not offering a constructor that takes a size. What do you think?

Reply:
Yeah... you don't really want an asynchronous constructor. 

But I don't like that if I want asynchronous allocation/free I can't use RAII. 

How about an API that returns a `std::future` to an instance of a `device_buffer` allocated on a stream? This still allows for asynchronous memory allocation/free that still uses RAII.

Reply:
Makes sense. Thanks for the clarification

Reply:
I get you point @jrhemstad regarding  `std::future`, however that does not change the issue of the constructor, right? It would replace the constructor with factory. Also you still have RAII when it is done like this
```cpp
template<typename T>
void foo( ML::cumlHandle* handle, .., cudaStream_t stream )
{
    ...
    ML::device_buffer<T> temp( handle->getDeviceAllocator(), 0 )
    
    temp.resize(n, stream);
    kernelA<<<grid, block, 0, stream>>>(..., temp.data(), ...);
    kernelB<<<grid, block, 0, stream>>>(..., temp.data(), ...);
    temp.release(stream);
}
```
in this case the pointer is still owned by temp. I.e. if the call to `release` does not happen for some reason (e.g. because a developer forgets to call it) the destructor of temp will make sure that the memory is released when temp goes out of scope. Does that make sense so you?

Reply:
I just pushed an update for device_buffer and host_buffer. Now the constructor takes a stream as an argument and the device_buffer and host_buffer are tied to that stream. The associated stream  a different stream can be changed by calling `resize`, `reserve´ or `release`.

Reply:
Nice. This now looks very aligned with our idea of not having to use default stream in cuml or ml-prims!